includes/Core/Consent_Mode/Consent_Mode.php

```diff
@@ -80,18 +80,64 @@ public function register() {
 	 * @since n.e.x.t
 	 */
 	protected function render_gtag_consent_snippet() {
-			$consent_defaults = array(
-				'ad_personalization' => 'denied',
-				'ad_storage'         => 'denied',
-				'ad_user_data'       => 'denied',
-				'analytics_storage'  => 'denied',
-				'regions'            => $this->consent_mode_settings->get_regions(),
-			);
-			?>
+		$consent_defaults = array(
+			'ad_personalization' => 'denied',
+			'ad_storage'         => 'denied',
+			'ad_user_data'       => 'denied',
+			'analytics_storage'  => 'denied',
+			'regions'            => $this->consent_mode_settings->get_regions(),
+			'wait_for_update'    => 500, // Allow 500ms for Consent Management Platforms (CMPs) to update the consent status.
+		);
+
+		$consent_category_map = apply_filters(
+			'googlesitekit_consent_category_map',
+			array(
+				'statistics' => array( 'analytics_storage' ),
+				'marketing'  => array( 'ad_storage', 'ad_user_data', 'ad_personalization' ),
+			)
+		);
+		// TODO: We should extract this JS and make it a bit more readable.
+		?>
 <!-- <?php echo esc_html__( 'Google tag (gtag.js) Consent Mode snippet added by Site Kit', 'google-site-kit' ); ?> -->
 <script id='google_gtagjs-js-consent-mode'>
 window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}
-gtag("consent","default", <?php echo wp_json_encode( $consent_defaults ); ?>);
+gtag("consent","default",<?php echo wp_json_encode( $consent_defaults ); ?>);
+window._googlesitekitConsentCategoryMap = <?php	echo wp_json_encode( $consent_category_map ); ?>;
+document.addEventListener('wp_listen_for_consent_change', function(event) {
+	if (event.detail) {
+		var consentParameters = Object.entries(event.detail).reduce((parameters,[category,status]) => {
+			if (window._googlesitekitConsentCategoryMap[category]) {
+				var mappedStatus = status === 'allow' ? 'granted' : 'denied';
+				window._googlesitekitConsentCategoryMap[category].forEach((parameter) => {
+					parameters[parameter] = mappedStatus;
+				});
+			}
+			return parameters;
+		}, {});
+		if (Object.keys(consentParameters).length) {
+			gtag('consent','update',consentParameters);
+		}
+	}
+});
+function updateGrantedConsent(){
+	if (!(window.wp_consent_type || window.wp_fallback_consent_type)){ return; }
```

Comment:
Bail out if neither `wp_consent_type` or `wp_fallback_consent_type` is set (i.e. a CMP plugin is not installed), otherwise we'll end up granting all consent parameters, as `wp_has_consent()` returns `true` if neither of these are set.